src/bgw/job_stat.c

```diff
@@ -0,0 +1,345 @@
+
+	return job_stat;
+}
+
+static bool
+bgw_job_stat_next_start_was_set(FormData_bgw_job_stat *fd)
+{
+	return fd->next_start != DT_NOBEGIN;
+}
+
+
+static bool
+bgw_job_stat_tuple_found(TupleInfo *ti, void *const data)
+{
+	BgwJobStat **job_stat_pp = data;
+
+	*job_stat_pp = bgw_job_stat_from_tuple(ti->tuple, ti->mctx);
+	return false;
+}
+
+static int
+bgw_job_stat_scan(int indexid, ScanKeyData scankey[], int nkeys,
+				  tuple_found_func tuple_found, tuple_filter_func tuple_filter, void *data, LOCKMODE lockmode)
+{
+	Catalog    *catalog = catalog_get();
+	ScannerCtx	scanCtx = {
+		.table = catalog->tables[BGW_JOB_STAT].id,
+		.index = CATALOG_INDEX(catalog, BGW_JOB_STAT, indexid),
+		.nkeys = nkeys,
+		.scankey = scankey,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+	};
+
+	return scanner_scan(&scanCtx);
+}
+
+static inline int
+bgw_job_stat_scan_job_id(int32 bgw_job_id, tuple_found_func tuple_found, tuple_filter_func tuple_filter, void *data, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_pkey_idx_job_id,
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(bgw_job_id));
+	return bgw_job_stat_scan(BGW_JOB_STAT_PKEY_IDX,
+							 scankey, 1, tuple_found, tuple_filter, data, lockmode);
+}
+
+BgwJobStat *
+ts_bgw_job_stat_find(int32 bgw_job_id)
+{
+	BgwJobStat *job_stat = NULL;
+
+	bgw_job_stat_scan_job_id(bgw_job_id, bgw_job_stat_tuple_found, NULL, &job_stat, AccessShareLock);
+
+	return job_stat;
+}
+
+
+static bool
+bgw_job_stat_tuple_mark_start(TupleInfo *ti, void *const data)
+{
+	HeapTuple	tuple = heap_copytuple(ti->tuple);
+	FormData_bgw_job_stat *fd = (FormData_bgw_job_stat *) GETSTRUCT(tuple);
+
+	fd->last_start = GetCurrentTimestamp();
+	fd->last_finish = DT_NOBEGIN;
+	fd->next_start = DT_NOBEGIN;
+
+	fd->total_runs++;
+
+	/*
+	 * This is udone by any of the end marks. This is so that we count crashes
+	 * conservatively. Pretty much the crash is incremented in the beginning
+	 * and then decremented during `bgw_job_stat_tuple_mark_end`. Thus, it
+	 * only remains incremented if the job is never marked as having ended.
+	 * This happens when: 1) the job crashes 2) another process crashes while
+	 * the job is running 3) the scheduler gets a SIGTERM while the job is
+	 * running
+	 *
+	 * Unfortunaetly, 3 cannot be helped because when a scheduler gets a
+	 * SIGTERM it sends SIGTERMS to it's any running jobs as well. Since you
+	 * aren't supposed to write to the DB Once you get a sigterm, neither the
+	 * job nor the scheduler can mark the end of a job.
+	 */
+	fd->last_run_success = false;
+	fd->total_crashes++;
+	fd->consecutive_crashes++;
+
+	catalog_update(ti->scanrel, tuple);
+	heap_freetuple(tuple);
+	return false;
+}
+
+
+typedef struct
+{
+	bool		success;
+	BgwJob	   *job;
+} JobResult;
+
+
+static
+TimestampTz
+calculate_next_start_success(TimestampTz last_finish, BgwJob *job)
```

Comment:
This function, and the next two, read like the they're calculating when the next successful/failing/crashing execution should run. Something like `calculate_next_start_on_success` might be clearer.

Reply:
That is better. changed.